### PR TITLE
Infrastructure for multiple memory pools

### PR DESCRIPTION
**Pull request by maleadt**
I created a couple of simple memory pools in order to understand the behavior/performance, but that has left me more confused than ever. Even with the pool we _currently_ have, I can't show any improvement over just plain CUDA malloc/free (@MikeInnes, do you remember which workload you initially created the pool for?). It's likely that CUDA malloc/free has improved, and/or the hardware matters (e.g. how much physical memory there is, whether or not is has hardware support for unified virtual addressing, etc).

Given the above, it's likely that we will need to support multiple types of pools. For example, no pool at all when CUDA malloc/free is shown to be fast enough, or a buddy allocator on CUDA 8.0/Kepler+ where we can actually implement a proper memory allocator by doing an overcommitted malloc of the entire device memory (mapping it into GPU memory space as we go).

This should be NFC, as it keeps the old pool, but includes some bugfixes I discovered along the way.

### CODE REVIEW

**Comment by maleadt** [comment]
Concrete example: @kristofferc's resnet50 from https://github.com/JuliaGPU/CuArrays.jl/issues/273 (tiny-imagenet-200, batch size 16, 32 batches, 1 epoch) on a Quadro RTX 5000 with 16GB of memory:

Straight CUDA malloc:

```
$ CUARRAYS_MEMORY_POOL=dummy jj --project resnet_julia_bench.jl
 ────────────────────────────────────────────────
                                   Time          
                           ──────────────────────
     Tot / % measured:          87.3s / 13.8%    

 Section           ncalls     time   %tot     avg
 ────────────────────────────────────────────────
 pooled alloc       54.6k    11.7s  97.2%   215μs
   1.1 alloc        54.6k    3.23s  26.8%  59.2μs
   2.0 gc(false)       56    8.19s  67.8%   146ms
     pooled free    46.1k    7.62s  63.2%   165μs
   2.1 alloc           56   43.9ms  0.36%   784μs
 pooled free        7.46k    343ms  2.84%  46.0μs
 ────────────────────────────────────────────────
Effective GPU memory usage: 10.52% (1.656 GiB/15.744 GiB)
CuArrays GPU memory usage: 659.567 MiB
DummyPool usage: 659.567 MiB (659.567 MiB allocated, 0 bytes cached)
DummyPool efficiency: 100.00% (659.567 MiB requested, 659.567 MiB allocated)
CuArrays.jl DummyPool statistics:
 - 54628 pool allocations: 295.555 GiB in 12.09s
 - 54628 CUDA allocations: 295.555 GiB in 11.13s
```

CUDA malloc not performing too bad, but this just seems like an insane amount of memory for this model? Correct me if I'm wrong here (cc @MikeInnes).

With the new splitting pool:

```
$ CUARRAYS_MEMORY_POOL=split jj --project resnet_julia_bench.jl
 ──────────────────────────────────────────────────────
                                         Time          
                                 ──────────────────────
        Tot / % measured:             85.8s / 12.4%    

 Section                 ncalls     time   %tot     avg
 ──────────────────────────────────────────────────────
 pooled alloc             54.6k    10.5s  98.6%   192μs
   1.1 scan               54.6k    775ms  7.30%  14.2μs
     pooled free              1   7.27μs  0.00%  7.27μs
   1.2 alloc              24.7k    2.44s  23.0%  99.0μs
   1.3 compact + scan        57   2.12ms  0.02%  37.2μs
   1.4 reclaim + alloc      109    5.81s  54.7%  53.3ms
   2.0 gc(false)             56    615ms  5.79%  11.0ms
     pooled free          46.2k   54.4ms  0.51%  1.18μs
   2.1 scan                  56   3.25ms  0.03%  58.0μs
 pooled free              7.45k    154ms  1.45%  20.6μs
 ──────────────────────────────────────────────────────
Effective GPU memory usage: 99.94% (15.735 GiB/15.744 GiB)
CuArrays GPU memory usage: 13.542 GiB
SplittingPool usage: 13.542 GiB (675.352 MiB allocated, 12.883 GiB cached)
SplittingPool efficiency: 4.76% (659.567 MiB requested, 13.542 GiB allocated)
CuArrays.jl SplittingPool statistics:
 - 54628 pool allocations: 295.555 GiB in 10.64s
 - 24595 CUDA allocations: 210.158 GiB in 8.05s
```

Some meager savings, but we can easily improve the implementation.

Now the pool we currently have:

```
$ CUARRAYS_MEMORY_POOL=binned jj --project resnet_julia_bench.jl
 ────────────────────────────────────────────────────
                                       Time          
                               ──────────────────────
       Tot / % measured:            94.6s / 18.4%    

 Section               ncalls     time   %tot     avg
 ────────────────────────────────────────────────────
 background task           19   74.8ms  0.43%  3.93ms
   pooled free          5.39k   1.44ms  0.01%   268ns
   reclaim                 19   22.7μs  0.00%  1.20μs
   scan                    19   14.1μs  0.00%   740ns
 pooled alloc           54.6k    17.3s  99.3%   316μs
   1. try alloc         9.05k    2.74s  15.8%   303μs
   2. gc(false)         1.62k    4.02s  23.1%  2.48ms
     pooled free        30.7k   9.87ms  0.06%   322ns
   3. reclaim unused    1.13k    2.13s  12.3%  1.89ms
     reclaim            1.13k    2.13s  12.3%  1.89ms
     scan               1.13k   1.03ms  0.01%   913ns
   4. try alloc         1.13k    5.08s  29.2%  4.50ms
   5. gc(true)             21    2.90s  16.7%   138ms
     pooled free        11.4k   2.84ms  0.02%   250ns
 pooled free            5.96k   44.9ms  0.26%  7.53μs
 ────────────────────────────────────────────────────
Effective GPU memory usage: 99.93% (15.733 GiB/15.744 GiB)
CuArrays GPU memory usage: 14.839 GiB
BinnedPool usage: 14.839 GiB (5.550 GiB allocated, 9.289 GiB cached)
BinnedPool efficiency: 27.88% (4.138 GiB requested, 14.839 GiB allocated)
CuArrays.jl BinnedPool statistics:
 - 54628 pool allocations: 295.555 GiB in 17.33s
 - 8531 CUDA allocations: 70.956 GiB in 7.94s
```

So here we have a significant saving on malloc calls, but it's not worth it since calling `gc(true)` (due to oversized allocations in the pool) is so much more expensive.

This effect is even more clear when limiting the memory usage to 8GB:

```
$ CUARRAYS_MEMORY_LIMIT=8000000000 CUARRAYS_MEMORY_POOL=dummy jj --project resnet_julia_bench.jl
 ────────────────────────────────────────────────
                                   Time          
                           ──────────────────────
     Tot / % measured:          92.0s / 17.9%    

 Section           ncalls     time   %tot     avg
 ────────────────────────────────────────────────
 pooled alloc       54.6k    16.3s  99.0%   299μs
   1.1 alloc        54.6k    2.81s  17.0%  51.4μs
   2.0 gc(false)      143    6.46s  39.1%  45.2ms
     pooled free    40.1k    5.86s  35.5%   146μs
   2.1 alloc          143   40.5ms  0.25%   283μs
   3.0 gc(true)        30    6.75s  40.9%   225ms
     pooled free    6.52k    2.57s  15.5%   394μs
   3.1 alloc           30   29.4ms  0.18%   980μs
 pooled free        6.70k    167ms  1.01%  24.9μs
 ────────────────────────────────────────────────
Effective GPU memory usage: 39.70% (6.250 GiB/15.744 GiB)
CuArrays GPU memory usage: 4.842 GiB (capped at 7.451 GiB)
DummyPool usage: 4.842 GiB (4.842 GiB allocated, 0 bytes cached)
DummyPool efficiency: 100.00% (4.842 GiB requested, 4.842 GiB allocated)
CuArrays.jl DummyPool statistics:
 - 54628 pool allocations: 295.555 GiB in 16.52s
 - 54628 CUDA allocations: 295.555 GiB in 11.41s
```

What's important to notice here: introduction of calls to `gc(true)`.

```
$ CUARRAYS_MEMORY_LIMIT=8000000000 CUARRAYS_MEMORY_POOL=split jj --project resnet_julia_bench.jl
 ────────────────────────────────────────────────────────────────────────
                                 Time                   Allocations      
                         ──────────────────────   ───────────────────────
    Tot / % measured:     18446737438s / 0.00%        15.8GiB / 35.5%    

 Section         ncalls     time   %tot     avg     alloc   %tot      avg
 ────────────────────────────────────────────────────────────────────────
 loss                32    20.1s  42.3%   630ms   2.18GiB  38.9%  69.9MiB
 back                32    19.4s  40.7%   605ms   2.36GiB  42.0%  75.4MiB
 update params       32    8.11s  17.0%   253ms   1.07GiB  19.1%  34.2MiB
 ────────────────────────────────────────────────────────────────────────
 ──────────────────────────────────────────────────────
                                         Time          
                                 ──────────────────────
        Tot / % measured:             90.2s / 15.0%    

 Section                 ncalls     time   %tot     avg
 ──────────────────────────────────────────────────────
 pooled alloc             54.6k    13.3s  98.9%   244μs
   1.1 scan               54.6k    395ms  2.93%  7.24μs
   1.2 alloc              22.8k    1.30s  9.67%  57.2μs
   1.3 compact + scan       181   3.07ms  0.02%  17.0μs
   1.4 reclaim + alloc      249    4.65s  34.5%  18.7ms
     pooled free             49   38.3μs  0.00%   782ns
   2.0 gc(false)            142    643ms  4.77%  4.53ms
     pooled free          40.5k   54.4ms  0.40%  1.34μs
   2.1 scan                 142   1.36ms  0.01%  9.59μs
   2.2 alloc                 54   15.3μs  0.00%   284ns
   2.3 compact + scan        33    366μs  0.00%  11.1μs
   2.4 reclaim + alloc       54    1.53s  11.3%  28.3ms
   3.0 gc(true)              29    3.98s  29.5%   137ms
     pooled free          6.12k   6.97ms  0.05%  1.14μs
   3.1 scan                  29    820μs  0.01%  28.3μs
 pooled free              6.80k    150ms  1.11%  22.0μs
 ──────────────────────────────────────────────────────
Effective GPU memory usage: 55.82% (8.788 GiB/15.744 GiB)
CuArrays GPU memory usage: 7.405 GiB (capped at 7.451 GiB)
SplittingPool usage: 7.405 GiB (3.390 GiB allocated, 4.015 GiB cached)
SplittingPool efficiency: 45.69% (3.384 GiB requested, 7.405 GiB allocated)
CuArrays.jl SplittingPool statistics:
 - 54628 pool allocations: 295.555 GiB in 13.5s
 - 22689 CUDA allocations: 164.512 GiB in 7.43s
```

New pool implementation very conservatively holds on to memory (the GPU memory pressure and thus number of calls to `gc(true)` and `gc(false)` doesn't increase), yet we have some nice savings.

```
$ CUARRAYS_MEMORY_LIMIT=8000000000 CUARRAYS_MEMORY_POOL=binned jj --project resnet_julia_bench.jl
 ────────────────────────────────────────────────────
                                       Time          
                               ──────────────────────
       Tot / % measured:             110s / 31.4%    

 Section               ncalls     time   %tot     avg
 ────────────────────────────────────────────────────
 background task           33    106ms  0.31%  3.21ms
   pooled free          2.70k    801μs  0.00%   297ns
   reclaim                 33   42.7μs  0.00%  1.29μs
   scan                    33   24.5μs  0.00%   743ns
 pooled alloc           54.6k    34.4s   100%   630μs
   1. try alloc         26.3k    1.12s  3.25%  42.8μs
   2. gc(false)         4.91k    10.8s  31.3%  2.21ms
     pooled free        27.8k   10.2ms  0.03%   368ns
   3. reclaim unused    3.81k    3.08s  8.91%   808μs
     reclaim            3.81k    3.07s  8.89%   806μs
     scan               3.81k   3.45ms  0.01%   904ns
   4. try alloc         3.81k    2.15s  6.20%   562μs
   5. gc(true)            125    16.9s  48.8%   135ms
     pooled free        17.3k   4.80ms  0.01%   277ns
 pooled free            5.61k   49.8ms  0.14%  8.88μs
 ────────────────────────────────────────────────────
Effective GPU memory usage: 52.92% (8.332 GiB/15.744 GiB)
CuArrays GPU memory usage: 7.437 GiB (capped at 7.451 GiB)
BinnedPool usage: 7.437 GiB (5.237 GiB allocated, 2.200 GiB cached)
BinnedPool efficiency: 52.42% (3.898 GiB requested, 7.437 GiB allocated)
CuArrays.jl BinnedPool statistics:
 - 54628 pool allocations: 295.555 GiB in 34.51s
 - 25044 CUDA allocations: 193.675 GiB in 6.32s
```

Ouch! The old pool was designed in an era / environment where CUDA malloc was much more expensive, and it was worth to hold on to much more memory at the expense if calling into the GC more often. This isn't the case on my system at all, hence the significant increase in total execution time.

TLDR: the cost of malloc has dropped, or changes depending on the hardware/CUDA toolkit, which doesn't make it worth for our memory pools to cache memory as aggressively. That said, the above benchmark allocates an insane amount of memory, and it's probably better to start optimizing that.

Kudos to @KristofferC for the nice timer infrastructure :smiley:

**Comment by vchuravy** [comment]
Nice!
